lib/Cache.js

```diff
@@ -7,54 +7,95 @@
 
 const { AsyncParallelHook, AsyncSeriesBailHook, SyncHook } = require("tapable");
 
-/** @typedef {(result: any, callback: (err?: Error) => void) => void} GotHandler */
+/** @typedef {import("./WebpackError")} WebpackError */
+
+/**
+ * @template T
+ * @callback Callback
+ * @param {WebpackError=} err
+ * @param {T=} stats
+ * @returns {void}
+ */
+
+/**
+ * @callback GotHandler
+ * @param {any} result
+ * @param {Callback<void>} stats
+ * @returns {void}
+ */
 
 class Cache {
 	constructor() {
 		this.hooks = {
-			/** @type {AsyncSeriesBailHook<string, string>} */
+			/** @type {AsyncSeriesBailHook<[string, string], any>} */
 			get: new AsyncSeriesBailHook(["identifier", "etag"]),
-			/** @type {AsyncParallelHook<string, string, any>} */
+			/** @type {AsyncParallelHook<[string, string, any]>} */
 			got: new AsyncParallelHook(["identifier", "etag", "data"]),
-			/** @type {AsyncParallelHook<string, string, any>} */
+			/** @type {AsyncParallelHook<[string, string, any]>} */
 			store: new AsyncParallelHook(["identifier", "etag", "data"]),
-			/** @type {SyncHook} */
+			/** @type {SyncHook<[]>} */
 			beginIdle: new SyncHook([]),
-			/** @type {AsyncParallelHook} */
+			/** @type {AsyncParallelHook<[]>} */
 			endIdle: new AsyncParallelHook([]),
-			/** @type {AsyncParallelHook} */
+			/** @type {AsyncParallelHook<[]>} */
 			shutdown: new AsyncParallelHook([])
 		};
 	}
 
+	/**
+	 * @template T
+	 * @param {string} identifier the cache identifier
+	 * @param {string} etag the etag
+	 * @param {Callback<T>} callback signals when the value is retrieved
+	 * @returns {void}
+	 */
 	get(identifier, etag, callback) {
 		this.hooks.get.callAsync(identifier, etag, (err, result) => {
 			if (err) {
-				callback(err);
+				callback(/** @type {WebpackError} */ (err));
```

Comment:
Assuming this is a WebpackError is incorrect. A plugin author is allowed to return any error.

We should create a HookWebpackError and wrap the error **if it's not a WebpackError**:

```suggestion
				callback(HookWebpackError.wrap(err, "Cache.hooks.get"));
```

Reply:
Should we do this in this PR or another one?

Reply:
In this PR or in a PR before this PR. Can't merge `callback(/** @type {WebpackError} */ (err));` as it's incorrect. I think it's ok to do it in this PR since the problem was discovered due to the new typings.